### PR TITLE
avm2: Add support for focus-related events

### PR DESCRIPTION
**Pull request by kjarosh**
Related to https://github.com/ruffle-rs/ruffle/issues/5443, progresses https://github.com/ruffle-rs/ruffle/issues/17084, fixes (probably) https://github.com/ruffle-rs/ruffle/issues/17064, progresses (probably) https://github.com/ruffle-rs/ruffle/issues/14993.

This PR adds support for focusing objects on mouse click, and for dispatching focus-related events:
* `focusOut`,
* `mouseFocusChange`, and
* `keyFocusChange`.

Behavior of those events and focusing objects by mouse is covered by tests:
* `avm2/focus_events_mouse_basic`,
* `avm2/focus_events_key_same_object`,
* `avm2/focus_events_key_basic`,
* `avm2/focus_events_code`,
* `avm2/focusrect_focuslost`,
* `avm1/focus_mouse_focusable`.

### CODE REVIEW

File: core/src/display_object/interactive.rs [hunk 1]
```diff
@@ -577,30 +576,8 @@ pub trait TInteractiveObject<'gc>:
             Avm1::run_stack_frame_for_method(self_do, object, context, method_name, &[other]);
         } else if let Avm2Value::Object(object) = self_do.object2() {
             let mut activation = Avm2Activation::from_nothing(context.reborrow());
-            let event_name = if focused {
-                "focusIn".into()
-            } else {
-                // `focusOut` is not this simple in FP,
-                // firing it might break SWFs that rely
-                // on the specific behavior
```
Comment (kjarosh) [inline core/src/display_object/interactive.rs:585]: @Lord-McSweeney Not sure what this meant, from my testing it seems that `focusOut` simply works that way

**Review by Dinnerbone — APPROVED** [review]
LGTM, thank you!

File: core/src/avm2/object/event_object.rs [hunk 2]
```diff
@@ -10,6 +10,7 @@ use crate::context::UpdateContext;
 use crate::display_object::TDisplayObject;
 use crate::display_object::{DisplayObject, InteractiveObject, TInteractiveObject};
 use crate::events::{KeyCode, MouseButton};
+use crate::prelude::Avm2Value;
```
Comment (Lord-McSweeney) [inline core/src/avm2/object/event_object.rs:13]: We already import `Value` here from `avm2::value::Value`; this is unnecessary

File: core/src/avm2/object/event_object.rs [hunk 3]
```diff
@@ -307,6 +308,38 @@ impl<'gc> EventObject<'gc> {
...
+        class
+            .construct(
+                activation,
+                &[
+                    event_type.into(),
+                    true.into(),
+                    cancelable.into(),
+                    related_object
+                        .map(|o| o.as_displayobject().object2())
+                        .unwrap_or(Avm2Value::Null),
```
Comment (Lord-McSweeney) [inline core/src/avm2/object/event_object.rs:335]: This can be `Value::Null` instead of `Avm2Value::Null`

**Review by Lord-McSweeney — APPROVED** [review]

File: core/src/avm2/object/event_object.rs (hunk 2 above)
Comment (kjarosh) [reply core/src/avm2/object/event_object.rs]: Right!

**Comment by torokati44** [comment]
Given two approvaps, and all review notes fixed, I guess this is good to go now...?

**Comment by GonzalezAtWork** [comment]
awesome work! thanks!

**Pull request merged**